test/built-ins/AsyncGeneratorPrototype/return/return-blocked-broken-promise.js

```diff
@@ -0,0 +1,59 @@
+// Copyright (C) 2022 Chengzhong Wu. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-asyncgenerator-prototype-return
+description: >
+  A broken promise should reject the returned promise of
+  AsyncGenerator.prototype.return when the generator is completed.
+info: |
+  AsyncGenerator.prototype.return ( value )
+  ...
+  8. If state is either suspendedStart or completed, then
+    a. Set generator.[[AsyncGeneratorState]] to awaiting-return.
+    b. Perform ! AsyncGeneratorAwaitReturn(generator).
+  ...
+
+  AsyncGeneratorAwaitReturn ( generator )
+  ...
+  6. Let promise be Completion(PromiseResolve(%Promise%, completion.[[Value]])).
+  7. If promiseCompletion is an abrupt completion, then
+    a. Set generator.[[AsyncGeneratorState]] to completed.
+    b. Perform AsyncGeneratorCompleteStep(generator, promiseCompletion, true).
+    c. Perform AsyncGeneratorDrainQueue(generator).
+    d. Return unused.
+  8. Assert: promiseCompletion.[[Type]] is normal.
+  9. Let promise be promiseCompletion.[[Value]].
+  ...
+
+flags: [async]
+features: [async-iteration]
+---*/
+
+let unblock;
+let blocking = new Promise(res => { unblock = res; });
```

Comment:
Although I believe I understand how this is working now, I found this difficult to comprehend on the first few readings :smile: One thing that got me on the wrong track was that I was reading `res` as "result" rather than "resolve" and was thinking of this arrow function as a callback passed to `then()` rather than the `(resolve, reject) => {...}` passed to the Promise constructor. Maybe writing out the name like this could make it clearer?
```suggestion
let blocking = new Promise(resolve => { unblock = resolve; });
```

Although, I'm not 100% clear on why we need to block the generator in order to test `return(brokenPromise)` in the "completed" state. Would it be sufficient to do something like
```js
var g = async function*() {};

var it = g();
it.return();
it.return(brokenPromise).then(...)
```
(There's probably a good reason why not, so maybe a comment would be helpful.)

Reply:
Ah, that explains it! Much clearer now with the new check, thank you.